kong/utils.go

```diff
@@ -249,33 +249,206 @@ func traverseConfigMap(currentConfigMap map[string]interface{}, path []string) (
+// linking deprecated field names to their corresponding replacement paths.
+//
+// The function iterates over the given `shorthandFields` JSON object, extracting the deprecated field names,
+// and their respective "replaced_with" or "translate_backwards" paths.
+// It builds a map where each key is a deprecated field name, and each value is,
+// a slice of replacement paths (each path itself is a slice of strings).
+//
+// Parameters:
+//
+//	shorthandFields (gjson.Result): A array of JSON objects containing shorthand field data.
+//																	Each shorthand field may contain deprecation information,
+//																	including the paths that should be used as replacements.
+//
+// Returns:
+//
+//	map[string][][]string: A map where each key is a deprecated field name (string),
+//	                        and each value is a slice of string slices representing the replacement paths for the field.
+//
+// Example:
+//
+//	shorthandFields := gjson.Parse(`[{
+//	    "field1": {
+//	        "deprecation": {
+//	            "replaced_with": [{"path": ["new_field1_1"]}, {"path": ["new_field1_2"]}]
+//	        }
+//	    },
+//	    "field2": {
+//	        "deprecation": {
+//	            "replaced_with": [{"path": ["new_field2"]}],
+//	        }
+//	    }
+//	}]`)
+//
+//	result := buildDeprecatedFieldWithReplacementsMap(shorthandFields)
+//	fmt.Println(result)
+//	// Output:
+//	// map[field1:[[new_field1_1] [new_field1_2]] field2:[[new_field2]]]
+func buildDeprecatedFieldWithReplacementsMap(shorthandFields gjson.Result) map[string][][]string {
+	result := make(map[string][][]string)
+	shorthandFields.ForEach(func(_, value gjson.Result) bool {
+		field := value.Map()
+		for deprecatedFieldName, shorthandFieldConfig := range field {
+			if paths := parseReplacedWithPaths(shorthandFieldConfig); paths != nil {
+				result[deprecatedFieldName] = paths
+			}
+		}
+
+		return true
+	})
+
+	return result
+}
+
+// processDeprecatedFields processes shorthand fields to build a map of deprecated field names
+// to their replacement paths, then filters out any shorthand fields not present in the given config.
+//
+// Parameters:
+//
+//	shorthandFields (gjson.Result): JSON object containing shorthand field data, including deprecation info.
+//	config (Configuration): A map that provides the valid shorthand field names to retain.
+//
+// Returns:
+//
+//	map[string][][]string: A map of deprecated field names to replacement paths, with invalid fields removed.
+func processDeprecatedFields(shorthandFields gjson.Result, config Configuration) map[string][][]string {
+	deprecatedFieldsWithReplacements := buildDeprecatedFieldWithReplacementsMap(shorthandFields)
+	filterValidShorthands(deprecatedFieldsWithReplacements, config)
+
+	return deprecatedFieldsWithReplacements
+}
+
+// filterValidShorthands removes any shorthand fields from the map that are not found in the given config.
```

Comment:
what does it mean when a shorthand field is in the map but not in the config? Could you give an example when/why this would be expected to happen?

Reply:
got it thanks for clarifying!